analyzers/src/SonarAnalyzer.CSharp/Rules/PrivateStaticMethodUsedOnlyByNestedClass.cs

```diff
@@ -0,0 +1,190 @@
+        type.Members
+                .OfType<MethodDeclarationSyntax>()
+                .Where(x => IsPrivateAndStatic(x, type))
+                .ToArray();
+
+    private static bool HasNestedTypeDeclarations(TypeDeclarationSyntax type) =>
+        type.Members
+                .OfType<TypeDeclarationSyntax>()
+                .Any();
+
+    private static bool IsPrivateAndStatic(MethodDeclarationSyntax method, TypeDeclarationSyntax containingType) =>
+        method.Modifiers.Any(x => x.IsKind(SyntaxKind.StaticKeyword))
+        && ((HasAnyModifier(method, SyntaxKind.PrivateKeyword) && !HasAnyModifier(method, SyntaxKind.ProtectedKeyword))
+           || (!HasAnyModifier(method, SyntaxKind.PublicKeyword, SyntaxKind.ProtectedKeyword, SyntaxKind.InternalKeyword) && IsClassOrRecordClassOrInterfaceDeclaration(containingType)));
+
+    private static bool IsPartial(TypeDeclarationSyntax type) =>
+        type.Modifiers.Any(x => x.IsKind(SyntaxKind.PartialKeyword));
+
+    private static bool IsClassOrRecordClassOrInterfaceDeclaration(TypeDeclarationSyntax type) =>
+        type is ClassDeclarationSyntax or InterfaceDeclarationSyntax
+        || (RecordDeclarationSyntaxWrapper.IsInstance(type) && !((RecordDeclarationSyntaxWrapper)type).ClassOrStructKeyword.IsKind(SyntaxKind.StructKeyword));
+
+    private static bool HasAnyModifier(MethodDeclarationSyntax method, params SyntaxKind[] modifiers) =>
+        method.Modifiers.Any(x => x.IsAnyKind(modifiers));
+
+    private static TypeDeclarationSyntax LowestCommonAncestorOrSelf(IEnumerable<TypeDeclarationSyntax> declaredTypes)
+    {
+        var treePaths = declaredTypes.Select(PathFromTop);
+        int minPathLength = treePaths.Select(x => x.Length).Min();
+        var firstPath = treePaths.First();
+
+        for (int i = 0; i < minPathLength; i++)
+        {
+            if (!treePaths.All(x => x[i] == firstPath[i]))
+            {
+                return firstPath[i - 1];
+            }
+        }
+
+        return firstPath[minPathLength - 1];
+
+        static TypeDeclarationSyntax[] PathFromTop(SyntaxNode node) =>
+            node.AncestorsAndSelf()
+                .OfType<TypeDeclarationSyntax>()
+                .Distinct()
+                .Reverse()
+                .ToArray();
+    }
+
+    private static IEnumerable<MethodUsedByTypes> TypesWhichUseTheMethods(
+        IEnumerable<MethodDeclarationSyntax> methods, TypeDeclarationSyntax outerType, SemanticModel model)
+    {
+        var collector = new PotentialMethodReferenceCollector(methods);
+        collector.Visit(outerType);
+
+        return collector.PotentialMethodReferences
+                            .Where(x => !OnlyUsedByOuterType(x))
+                            .Select(ResolveReferences)
+                            .Where(x => x.Types.Any())
+                            .ToArray();
+
+        MethodUsedByTypes ResolveReferences(MethodWithPotentialReferences m)
```

Comment:
this is getting the Type of the reference, right? "Resolve" is too vague for a name - it could be symbol resolution etc

Reply:
Renamed it to `DeclaredTypesWhichActuallyUseTheMethod()`.